backend/src/index.ts

```diff
@@ -15,11 +15,46 @@ import * as archives from './config/archives.json';
 import * as sourceChains from './config/sourceChains.json';
 import { getChainName } from './httpApi';
 
+type Task = () => Promise<void>;
+
+class AsyncQueue {
+  private readonly concurrency: number;
+  private pendingCount: number;
+  private readonly waiting: Task[];
+
+  constructor(concurrency: number) {
+    this.concurrency = concurrency;
+    this.pendingCount = 0;
+    this.waiting = [];
+  }
+
+  add(task: Task) {
+    const isWithinLimit = this.pendingCount < this.concurrency;
+    if (isWithinLimit) {
+      this.next(task);
+      return;
+    }
+    this.waiting.push(task);
+  }
+
+  async next(task: Task) {
+    this.pendingCount++;
+    await task();
+    this.pendingCount--;
+
+    if (this.waiting.length > 0) {
+      const task = this.waiting.shift();
+      this.next(task as Task);
+      return;
+    }
```

Comment:
No need to convince type system in something it knows for sure:
```suggestion
    const task = this.waiting.shift();
    if (task) {
      this.next(task);
      return;
    }
```